tests/host_tools/cargo_build.py

```diff
@@ -55,24 +55,29 @@ def get_firecracker_binaries():
     Returns the location of the firecracker related binaries eventually after
     building them in case they do not exist at the specified root_path.
     """
+    if not hasattr(get_firecracker_binaries, "was_built"):
```

Comment:
If any source file changes and we do `./tools/devtool build`, will the build get trigger again?

Reply:
yes. this attribute lives as long as the script execution. when we do a new `tools/devtool build`, the build will get triggered again.

This was needed because until now, this function did a `cargo build` on every call. However, since we now update the `build.rs` on every build, this was taking too long and some tests were failing with timeout.

Reply:
The thing I don't like about this is: with this PR _every time_ I run `pytest` or `devtool test` when doing development, it is going to do a top-level `release` build. That time can really add up during a development session where you run a test repeatedly.

I'd like to have a way to either:
- switch this logic off (based on env variable toggled from `devtool`: **on** for `build`, **off** for everything else `test/shell/etc`)
- use something mechanism independent of `cargo`/`build.rs` so that we get finer control over when it is enabled or not (for example we want it in `devtool build` but not in `devtool test`)
- some other idea to avoid rebuilding each time we instantiate `pytest`

Reply:
With this PR, cargo is only going to recompile and rerun the build script, since nothing else is changed in the source files.

I have run the `time tools/devtool build` command to see how much it affects build time and it is only `2.785s`.

Is this really a concern?

Reply:
Can you do a quick profile of the `get_firecracker_binaries()` function when running `pytest <some-test>` with and without the rebuild? Ideally not new vs old code as old code had unnecessary `strip`.

I would like to compare two runs where the code hasn't changed (a rebuild should not trigger) (without `strip`):
1. with the `touch build.rs` which will force re-trigger build
2. without `touch build.rs` which will be basically a no-op

If the difference is 1-2 seconds, then I am fine with it, but I was under the impression that the binary is _always_ rebuilt (including linking) which on `--release` (such as tests) is considerable.

Reply:
So I've done some testing:

This is how this PR behaves in its current form (where `touch src/firecracker/build.rs` always happens):

```
acatan@uf6ed9c851f4556:~/code/firecracker$ git log -2 --oneline --decorate
321433b (HEAD -> improve_version, alin/improve_version) Merge branch 'master' into improve_version
a2a6822 Improve firecracker --version output. Update build script to build the binaries only once.

acatan@uf6ed9c851f4556:~/code/firecracker$ cargo build --release
Updating crates.io index
  Downloaded versionize v0.1.2
   Compiling libc v0.2.68
...
   Compiling api_server v0.1.0 (/home/ANT.AMAZON.COM/acatan/code/firecracker/src/api_server)
    Finished release [optimized] target(s) in 1m 36s

acatan@uf6ed9c851f4556:~/code/firecracker$ cargo build --release
   Compiling firecracker v0.21.0 (/home/ANT.AMAZON.COM/acatan/code/firecracker/src/firecracker)
    Finished release [optimized] target(s) in 20.70s

acatan@uf6ed9c851f4556:~/code/firecracker$ cargo build --release
   Compiling firecracker v0.21.0 (/home/ANT.AMAZON.COM/acatan/code/firecracker/src/firecracker)
    Finished release [optimized] target(s) in 20.74s
```

**Without changing any files** in the repository, each `build --release` **takes 20.74s** on my local machine (which is faster than a EC2 m5d.metal).

Removing the forced `touch src/firecracker/build.rs`:

```
acatan@uf6ed9c851f4556:~/code/firecracker$ git diff
diff --git a/src/firecracker/build.rs b/src/firecracker/build.rs
index 83cbeec..7bb283c 100644
--- a/src/firecracker/build.rs
+++ b/src/firecracker/build.rs
@@ -27,5 +27,4 @@ fn main() {
         );
     }
 
-    Command::new("touch").args(&["build.rs"]).output().unwrap();
 }

acatan@uf6ed9c851f4556:~/code/firecracker$ cargo build --release
   Compiling firecracker v0.21.0 (/home/ANT.AMAZON.COM/acatan/code/firecracker/src/firecracker)
    Finished release [optimized] target(s) in 20.68s

acatan@uf6ed9c851f4556:~/code/firecracker$ cargo build --release
    Finished release [optimized] target(s) in 0.05s

acatan@uf6ed9c851f4556:~/code/firecracker$ cargo build --release
    Finished release [optimized] target(s) in 0.04s
```

This is correct behavior, nothing changed, iterative `cargo build --release` take **0.04s**.

Reply:
Indeed. I was only taking into account the `devtool build` command, without the `--release` flag.
I will have to update the implementation to account for this. 20 seconds is not acceptable.

I will adapt the implementation to do a `touch build.rs` only from `devtool/build --release`